From d6338b308affcca8a0021e95341fd1ed5563e5db Mon Sep 17 00:00:00 2001
From: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Date: Fri, 28 Aug 2026 02:23:49 +0800
Subject: [PATCH 3/4] Explicitly load the ZFS module via systemd service

Add zfs-load-module.service to explicitly load the ZFS kernel module
before importing pools.

AOSC: Refresh the change against the current upstream systemd units.
---
 contrib/dracut/90zfs/module-setup.sh.in        |  3 ++-
 etc/Makefile.am                                |  1 +
 etc/systemd/system/50-zfs.preset               |  1 +
 etc/systemd/system/zfs-import-cache.service.in |  2 ++
 etc/systemd/system/zfs-import-scan.service.in  |  2 ++
 etc/systemd/system/zfs-load-module.service.in  | 17 +++++++++++++++++
 6 files changed, 25 insertions(+), 1 deletion(-)
 create mode 100644 etc/systemd/system/zfs-load-module.service.in

diff --git a/contrib/dracut/90zfs/module-setup.sh.in b/contrib/dracut/90zfs/module-setup.sh.in
index 130d94c70..f765cb26c 100755
--- a/contrib/dracut/90zfs/module-setup.sh.in
+++ b/contrib/dracut/90zfs/module-setup.sh.in
@@ -96,7 +96,8 @@ install() {
 
 		for _service in \
 			"zfs-import-scan.service" \
-			"zfs-import-cache.service"; do
+			"zfs-import-cache.service" \
+			"zfs-load-module.service"; do
 			inst_simple "${systemdsystemunitdir}/${_service}"
 			systemctl -q --root "${initdir}" add-wants zfs-import.target "${_service}"
 
diff --git a/etc/Makefile.am b/etc/Makefile.am
index a795bec16..cafd51ebb 100644
--- a/etc/Makefile.am
+++ b/etc/Makefile.am
@@ -47,6 +47,7 @@ dist_systemdpreset_DATA = \
 	%D%/systemd/system/50-zfs.preset
 
 systemdunit_DATA = \
+	%D%/systemd/system/zfs-load-module.service \
 	%D%/systemd/system/zfs-import-cache.service \
 	%D%/systemd/system/zfs-import-scan.service \
 	%D%/systemd/system/zfs-import.target \
diff --git a/etc/systemd/system/50-zfs.preset b/etc/systemd/system/50-zfs.preset
index e4056a92c..b53e2c8a8 100644
--- a/etc/systemd/system/50-zfs.preset
+++ b/etc/systemd/system/50-zfs.preset
@@ -7,3 +7,4 @@ enable zfs-share.service
 enable zfs-zed.service
 enable zfs-volume-wait.service
 enable zfs.target
+enable zfs-load-module.service
diff --git a/etc/systemd/system/zfs-import-cache.service.in b/etc/systemd/system/zfs-import-cache.service.in
index c7aa34ccb..a86695dec 100644
--- a/etc/systemd/system/zfs-import-cache.service.in
+++ b/etc/systemd/system/zfs-import-cache.service.in
@@ -3,7 +3,9 @@ Description=Import ZFS pools by cache file
 Documentation=man:zpool(8)
 DefaultDependencies=no
 Wants=systemd-udev-settle.service
+Requires=zfs-load-module.service
 After=systemd-udev-settle.service
+After=zfs-load-module.service
 After=systemd-udev-trigger.service
 After=systemd-modules-load.service
 After=cryptsetup.target
diff --git a/etc/systemd/system/zfs-import-scan.service.in b/etc/systemd/system/zfs-import-scan.service.in
index b83476771..ac20a77e2 100644
--- a/etc/systemd/system/zfs-import-scan.service.in
+++ b/etc/systemd/system/zfs-import-scan.service.in
@@ -3,7 +3,9 @@ Description=Import ZFS pools by device scanning
 Documentation=man:zpool(8)
 DefaultDependencies=no
 Wants=systemd-udev-settle.service
+Requires=zfs-load-module.service
 After=systemd-udev-settle.service
+After=zfs-load-module.service
 After=systemd-udev-trigger.service
 After=systemd-modules-load.service
 After=cryptsetup.target
diff --git a/etc/systemd/system/zfs-load-module.service.in b/etc/systemd/system/zfs-load-module.service.in
new file mode 100644
index 000000000..d74af0468
--- /dev/null
+++ b/etc/systemd/system/zfs-load-module.service.in
@@ -0,0 +1,17 @@
+[Unit]
+Description=Install ZFS kernel module
+DefaultDependencies=no
+Requires=systemd-udev-settle.service
+After=systemd-udev-settle.service
+After=cryptsetup.target
+Before=dracut-mount.service
+After=systemd-remount-fs.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart=/sbin/modprobe zfs
+
+[Install]
+WantedBy=zfs-mount.service
+WantedBy=zfs.target
-- 
2.55.0

